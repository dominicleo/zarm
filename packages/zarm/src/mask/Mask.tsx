import * as React from 'react';
import { createBEM } from '@zarm-design/bem';
import { CSSTransition } from 'react-transition-group';
import { ConfigContext } from '../n-config-provider';
import type { HTMLProps, NativeAttrs } from '../utils/utilityTypes';
import type { BaseMaskProps } from './interface';

export interface MaskCssVars {
  '--background-color'?: React.CSSProperties['backgroundColor'];
  '--zindex'?: React.CSSProperties['zIndex'];
  '--filter'?: string;
}

export type MaskProps = NativeAttrs<
  React.HTMLAttributes<HTMLDivElement>,
  BaseMaskProps & HTMLProps<MaskCssVars>
>;

const Mask = React.forwardRef<HTMLDivElement, MaskProps>((props, ref) => {
  const { className, visible = false, blur, duration = 300, children, ...rest } = props;
  const { prefixCls } = React.useContext(ConfigContext);
  const bem = createBEM('mask', { prefixCls });

  return (
    <CSSTransition
      mountOnEnter
      unmountOnExit
      in={visible}
      timeout={duration}
      classNames={`${prefixCls}-fade`}
    >
      <div {...rest} ref={ref} className={bem([{ blur }, className])}>
        {children}
      </div>
    </CSSTransition>
  );
});

export default Mask;
