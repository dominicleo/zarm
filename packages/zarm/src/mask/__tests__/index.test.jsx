import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import Mask from '../index';

describe('Mask', () => {
  it('renders correctly', () => {
    const wrapper = render(<Mask visible />);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('blur method', () => {
    const wrapper = render(<Mask visible blur />);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('onClick', async () => {
    const onClick = jest.fn();
    render(<Mask data-testid='mask' visible onClick={onClick} />);
    const mask = screen.getByTestId('mask');
    fireEvent.click(mask);
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
