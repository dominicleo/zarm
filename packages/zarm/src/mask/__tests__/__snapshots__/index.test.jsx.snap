// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mask blur method 1`] = `
<DocumentFragment>
  <div
    class="za-mask za-mask--blur"
  />
</DocumentFragment>
`;

exports[`Mask renders correctly 1`] = `
<DocumentFragment>
  <div
    class="za-mask"
  />
</DocumentFragment>
`;
